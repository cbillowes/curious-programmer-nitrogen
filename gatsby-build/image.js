const _ = require("lodash")
const path = require(`path`)
const fs = require("fs")
const sharp = require("sharp")

const rawSrc = `src/images/raw`
const rawDest = `src/images/root`
const rawQuality = 100

const imgTemplateSrc = path.join(__dirname, `img.jsx.template`)
const socialMediaQuality = 80
const socialMediaSrc = path.join(__dirname, `../src/images/root`)
const socialMediaRootDest = path.join(__dirname, `../src/images/social-media`)
const gifsDest = path.join(__dirname, `../`, `/public/gifs`)
const svgsDest = path.join(__dirname, `../`, `/public/svgs`)
const socialMediaDest = path.join(__dirname, `../`, `/public/social-media`)
const staticDest = path.join(__dirname, `../public`)
const componentsDest = path.join(__dirname, `../`, `/src/components/images/`)

/**
 * Keeps a copy of an optimized original image that has been resized to ensure uniformity.
 * @param {string} src
 */
const toRoot = src => {
  const dest = getDest(src, rawDest)
  const quality = rawQuality

  mkdir(rawDest)
  sharp(src)
    .jpeg({ quality })
    .png({ quality })
    .webp({ quality })
    .resize(1920, 1080, {
      fit: sharp.fit.cover,
      position: sharp.strategy.attention,
    })
    .toFile(dest, err => {
      const message = `optimize image: ${src} -> ${dest}`
      if (err) {
        return `${message}\n${err}`
      } else {
        return message
      }
    })
}

/**
 * Creates an optimized image resized for social media sharing.
 * @param {string} src
 */
const toSocialMedia = src => {
  const dest = getDest(src, socialMediaRootDest)
  const quality = socialMediaQuality
  const dimensions = { width: 1200, height: 600 }
  mkdir(socialMediaRootDest)
  sharp(src)
    .jpeg({ quality })
    .png({ quality })
    .webp({ quality })
    .resize({
      width: dimensions.width,
      height: dimensions.height,
      kernel: sharp.kernel.nearest,
      fit: sharp.fit.cover,
    })
    .toFile(dest, err => {
      const message = `social media image: ${src} -> ${dest}`
      if (err) {
        return `${message}\n${err}`
      } else {
        return Promise.resolve(message)
      }
    })
}

const getComponentName = filename => {
  const extensionless = filename.replace(path.extname(filename), ``)
  const camel = _.camelCase(extensionless)
  const upper = _.startCase(camel)
  const tidied = upper.replace(/ /g, ``).replace(/\-/g, ``)
  return tidied
}

const toComponent = src => {
  mkdir(componentsDest)
  fs.readFile(imgTemplateSrc, (err, data) => {
    if (err) {
      return `image template: ${src}\n${err}`
    } else {
      const filename = path.basename(src)
      const destFilename = filename.replace(path.extname(filename), `.js`)
      const dest = path.join(componentsDest, destFilename)

      const componentName = getComponentName(filename)
      const imageFilename = filename
      const imageWidth = 960

      const component = data
        .toString()
        .replace(/%COMPONENT_NAME%/g, componentName)
        .replace(/%IMAGE_FILENAME%/g, imageFilename)
        .replace(/%IMAGE_WIDTH%/, imageWidth)

      fs.writeFile(dest, component, () => {
        return `component: -> ${dest}`
      })
    }
  })
}

/**
 * Deletes the raw image to save space. This way hi-res images can be processed and
 * large media does not need to be stored in this git repo.
 * @param {string} src
 * @param {object} reporter
 */
const unlink = async src => {
  fs.unlink(src, err => {
    if (err) {
      throw new Error(`unlink image: ${src} -> ${dest}`)
    }
  })
}

module.exports.process = (src, reporter) => {
  if (src.indexOf(rawSrc) >= 0) {
    const fns = [toRoot, toSocialMedia, toComponent]
    let promise = Promise.resolve()

    fns.forEach(fn => {
      promise = promise.then(() => {
        reporter.verbose(`process`)
        return fn(src)
      })
    })
    promise.finally(() => unlink(src))
  }
}

module.exports.toStatic = (src, mediaType, reporter) => {
  if (
    mediaType === `image/gif` ||
    mediaType === `image/svg+xml` ||
    src.indexOf(socialMediaRootDest) > -1
  ) {
    const destPath = getStaticDirectory(mediaType, src)
    const dest = path.join(destPath, path.basename(src))

    mkdir(destPath)
    fs.copyFile(src, dest, err => {
      const message = `${mediaType}: ${src} -> ${dest}`
      if (err) {
        reporter.error(`${message}\n${err}`)
      } else {
        reporter.verbose(message)
      }
    })
  }
}

const getStaticDirectory = (mediaType, src) => {
  if (mediaType === `image/gif`) return gifsDest
  if (mediaType === `img/svg+xml`) return svgsDest
  if (src.indexOf(socialMediaRootDest) > -1) return socialMediaDest
  return staticDest
}

const mkdir = path => {
  fs.existsSync(path) || fs.mkdirSync(path, { recursive: true })
}

const getDest = (src, dest) => {
  return path.join(dest, path.basename(src))
}

/**
 * Synchronize your asynchronous code using JavaScriptâ€™s async await
 * Patrick Ferreira
 * https://medium.com/@patarkf/synchronize-your-asynchronous-code-using-javascripts-async-await-5f3fa5b1366d
 */

//TODO
// Find out why svg is blank on build
// Use error message function
// Use Promise.all
