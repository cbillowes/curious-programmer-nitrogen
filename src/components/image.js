import React from "react"
import { useStaticQuery, graphql } from "gatsby"
import Img from "gatsby-image"

function render(file) {
  return <Img fluid={file.node.childImageSharp.fluid} />
}

function getImage(files, src) {
  return files.images.edges.find(image => image.node.relativePath === src)
}

const Image = ({ src }) => {
  const images = useStaticQuery(graphql`
    query {
      images: allFile(filter: { extension: { regex: "/jpeg|jpg|png|gif/"}}) {
        edges {
          node {
            extension
            relativePath
            childImageSharp {
              fluid(maxWidth: 1980) {
                ...GatsbyImageSharpFluid
              }
            }
          }
        }
      }
    }
  `)
  const image = getImage(images, src)
  return render(image)
}

export default Image
