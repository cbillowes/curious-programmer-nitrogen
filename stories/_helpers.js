import React from 'react'
import Layout from '../src/components/layout'
import { SmallSpacer } from '../src/components/spacer'

const Page = ({ className, children }) => {
  return (
    <Layout
      className={className}
    >
      <SmallSpacer />
      {children}
    </Layout>
  )
}

export default Page