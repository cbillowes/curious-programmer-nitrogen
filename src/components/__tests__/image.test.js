import React from 'react'
import { PureImage as Image } from '../image'
import { getSnapshot, mustContainValue, mustNotContainValue } from './_helpers'

const data = {
  "images": {
    "edges": [
      {
        "node": {
          "extension": "png",
          "relativePath": "favicon.png",
          "childImageSharp": {
            "fluid": {
              "sizes": "(max-width: 64px) 100vw, 64px",
              "src": "/static/902e86a7b3f31818f9cd33b4adb71503/fc3a2/favicon.png",
              "srcSet": "/static/902e86a7b3f31818f9cd33b4adb71503/fc3a2/favicon.png 64w",
              "aspectRatio": 1
            }
          }
        }
      },
      {
        "node": {
          "extension": "png",
          "relativePath": "gatsby-astronaut.png",
          "childImageSharp": {
            "fluid": {
              "sizes": "(max-width: 800px) 100vw, 800px",
              "src": "/static/6d91c86c0fde632ba4cd01062fd9ccfa/a54c6/gatsby-astronaut.png",
              "srcSet": "/static/6d91c86c0fde632ba4cd01062fd9ccfa/59beb/gatsby-astronaut.png 200w,\n/static/6d91c86c0fde632ba4cd01062fd9ccfa/c0bbd/gatsby-astronaut.png 400w,\n/static/6d91c86c0fde632ba4cd01062fd9ccfa/a54c6/gatsby-astronaut.png 800w",
              "aspectRatio": 1
            }
          }
        }
      },
    ]
  }
}

function ImageUnderTest(props) {
  return (
    <Image
      data={data}
      src={props.src}
    />
  )
}

describe(`Image`, () => {
  it(`should render an image that exists`, () => {
    const image = (
      <Image
        data={data}
        src="favicon.png"
      />
    )
    const tree = getSnapshot(image)
    expect(tree).toMatchSnapshot()
  })

  it(`should not render an image that does not exist`, () => {
    const image = (
      <Image
        data={data}
        src="i-do-not-exist.png"
      />
    )
    const tree = getSnapshot(image)
    expect(tree).toMatchSnapshot()
  })
})
