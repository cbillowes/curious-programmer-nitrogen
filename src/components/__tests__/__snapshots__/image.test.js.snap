// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image should not render an image that does not exist 1`] = `null`;

exports[`Image should render an image that exists 1`] = `
<div
  className=" gatsby-image-wrapper"
  style={
    Object {
      "overflow": "hidden",
      "position": "relative",
    }
  }
>
  <div
    aria-hidden={true}
    style={
      Object {
        "paddingBottom": "100%",
        "width": "100%",
      }
    }
  />
  <picture>
    <source
      sizes="(max-width: 64px) 100vw, 64px"
      srcSet="/static/902e86a7b3f31818f9cd33b4adb71503/fc3a2/favicon.png 64w"
    />
    <img
      alt=""
      loading="lazy"
      onLoad={[Function]}
      sizes="(max-width: 64px) 100vw, 64px"
      src="/static/902e86a7b3f31818f9cd33b4adb71503/fc3a2/favicon.png"
      srcSet="/static/902e86a7b3f31818f9cd33b4adb71503/fc3a2/favicon.png 64w"
      style={
        Object {
          "height": "100%",
          "left": 0,
          "objectFit": "cover",
          "objectPosition": "center",
          "opacity": 0,
          "position": "absolute",
          "top": 0,
          "transition": "opacity 500ms",
          "width": "100%",
        }
      }
    />
  </picture>
  <noscript
    dangerouslySetInnerHTML={
      Object {
        "__html": "<picture><source srcset=\\"/static/902e86a7b3f31818f9cd33b4adb71503/fc3a2/favicon.png 64w\\" sizes=\\"(max-width: 64px) 100vw, 64px\\" /><img loading=\\"lazy\\" sizes=\\"(max-width: 64px) 100vw, 64px\\" srcset=\\"/static/902e86a7b3f31818f9cd33b4adb71503/fc3a2/favicon.png 64w\\" src=\\"/static/902e86a7b3f31818f9cd33b4adb71503/fc3a2/favicon.png\\" alt=\\"\\" style=\\"position:absolute;top:0;left:0;opacity:1;width:100%;height:100%;object-fit:cover;object-position:center\\"/></picture>",
      }
    }
  />
</div>
`;
